src/input/polyanya_file.rs

```diff
@@ -16,69 +19,67 @@ pub struct PolyanyaFile {
 }
 
 impl PolyanyaFile {
-    /// Create a `Mesh` from a file in the format `mesh 2`.
+    /// Create a `Mesh` from a file in the formats `mesh 2` or `mesh 3`.
     ///
-    /// See <https://github.com/vleue/polyanya/blob/main/meshes/format.txt> for format description.
+    /// See <https://github.com/vleue/polyanya/blob/main/meshes/v2/format.txt> for v2 format description, or
+    /// <https://github.com/vleue/polyanya/blob/main/meshes/v3/format.txt>
     pub fn from_file(path: &str) -> PolyanyaFile {
         let mut file = std::fs::File::open(path).unwrap();
         let mut buffer = Vec::new();
         file.read_to_end(&mut buffer).unwrap();
         Self::from_bytes(&buffer)
     }
 
-    /// Create a `Mesh` from bytes in the format `mesh 2`.
+    /// Create a `Mesh` from bytes in the formats `mesh 2` or `mesh 3`.
     ///
-    /// See <https://github.com/vleue/polyanya/blob/main/meshes/format.txt> for format description.
+    /// See <https://github.com/vleue/polyanya/blob/main/meshes/v2/format.txt> for v2 format description, or
+    /// <https://github.com/vleue/polyanya/blob/main/meshes/v3/format.txt>
     pub fn from_bytes(bytes: &[u8]) -> PolyanyaFile {
-        let mut mesh = PolyanyaFile {
-            vertices: vec![],
-            polygons: vec![],
-        };
-        let mut nb_vertices = 0;
-        let mut nb_polygons = 0;
-        let mut phase = 0;
-        for line in io::BufReader::new(bytes).lines() {
-            let line: String = line.unwrap();
-            if phase == 0 {
-                if line == "mesh" || line == "2" {
-                    continue;
-                } else {
-                    (nb_vertices, nb_polygons) = line
-                        .split_once(' ')
-                        .map(|(a, b)| (a.parse().unwrap(), b.parse().unwrap()))
-                        .unwrap();
-                    phase = 1;
-                    continue;
-                }
-            }
-            if phase == 1 {
-                if nb_vertices > 0 {
-                    nb_vertices -= 1;
-                    let mut values = line.split(' ');
-                    let x = values.next().unwrap().parse().unwrap();
-                    let y = values.next().unwrap().parse().unwrap();
-                    let _ = values.next();
-                    let vertex = Vertex::new(
-                        Vec2::new(x, y),
-                        values.map(|v| v.parse().unwrap_or(u32::MAX)).collect(),
-                    );
-                    mesh.vertices.push(vertex);
-                } else {
-                    phase = 2;
-                }
+        let mut lines = BufReader::new(bytes).lines();
+
+        // First check the header.
+        if let Some(Ok(first)) = lines.next() {
+            if first != "mesh" {
+                panic!("Invalid polyanya .mesh file");
+            } else {
+                println!("mesh");
```

Comment:
debug println? 🙂

Reply:
yup lol

Reply:
I don't really like how I've implemented header validation here, so if you have any better ideas let me know.

Also the duplicated `panic!`s suck (tbh I'd rather it didn't panic, but didn't want to overhaul it to return a result and bubble that up), but let chains aren't stable yet (but will be in the next rust version 1.88).

Reply:
ok for me... it can be refactored later if that still bothers you 😄